clang/runtime/dpct-rt/include/dpl_extras/vector.h.inc

```diff
@@ -121,143 +122,142 @@ private:
-                                      ::std::random_access_iterator_tag>::value,
-                    InputIterator>::type last)
-      : _alloc(get_default_queue()), _size(::std::distance(first, last)) {
-    _set_capacity_and_alloc();
-    ::std::vector<T> _tmp(first, last);
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                _tmp.begin(), _tmp.end(), this->begin());
-    }
+  device_vector(const device_vector &other)
+      : _alloc(std::allocator_traits<Allocator>::
+                   select_on_container_copy_construction(other._alloc)) {
+    _size = other.size();
+    _capacity = other.capacity();
+    _storage = ::std::allocator_traits<Allocator>::allocate(_alloc, _capacity);
+    _construct(other.begin(), other.end());
   }
 
   template <typename OtherAllocator>
-  device_vector(const device_vector<T, OtherAllocator> &v)
-      : _alloc(get_default_queue()), _storage(v.real_begin()), _size(v.size()),
-        _capacity(v.capacity()) {}
+  device_vector(const device_vector<T, OtherAllocator> &other)
+      : _alloc(get_default_queue()) {
+    _size = other.size();
+    _capacity = other.capacity();
+    _storage = ::std::allocator_traits<Allocator>::allocate(_alloc, _capacity);
+    _construct(other.begin(), other.end());
+  }
 
   template <typename OtherAllocator>
   device_vector(::std::vector<T, OtherAllocator> &v)
-      : _alloc(get_default_queue()), _size(v.size()) {
-    _set_capacity_and_alloc();
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                v.begin(), v.end(), this->begin());
-    }
-  }
-
+      : device_vector(v.begin(), v.end()) {}
   template <typename OtherAllocator>
   device_vector &operator=(const ::std::vector<T, OtherAllocator> &v) {
     resize(v.size());
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                v.begin(), v.end(), begin());
-    }
+    _construct(v.begin(), v.end());
     return *this;
   }
   device_vector &operator=(const device_vector &other) {
     // Copy assignment operator:
-    resize(other.size());
-    if (_size > 0) {
-      ::std::copy(oneapi::dpl::execution::make_device_policy(get_default_queue()),
-                other.begin(), other.end(), begin());
+    if constexpr (::std::allocator_traits<Allocator>::
+                      propagate_on_container_copy_assignment::value) {
+      _alloc = other._alloc;
     }
+    resize(other.size());
+    _construct(other.begin(), other.end());
     return *this;
   }
   device_vector &operator=(device_vector &&other) {
     // Move assignment operator:
-    device_vector dummy(::std::move(other));
-    this->swap(dummy);
+    if constexpr (::std::allocator_traits<Allocator>::
+                      propagate_on_container_move_assignment::value) {
+      _alloc = std::move(other._alloc);
```

Comment:
Should we deallocate current storage here?

Reply:
Yes, done.